dbt/adapters/spark/connections.py

```diff
@@ -103,7 +121,51 @@ def execute(self, sql, bindings=None):
         if sql.strip().endswith(";"):
             sql = sql.strip()[:-1]
 
-        return self._cursor.execute(sql, bindings)
+        # Reaching into the private enumeration here is bad form,
+        # but there doesn't appear to be any way to determine that
+        # a query has completed executing from the pyhive public API.
+        # We need to use an async query + poll here, otherwise our
+        # request may be dropped after ~5 minutes by the thrift server
+        STATE_PENDING = [
+            ThriftState.INITIALIZED_STATE,
+            ThriftState.RUNNING_STATE,
+            ThriftState.PENDING_STATE,
+        ]
+
+        STATE_SUCCESS = [
+            ThriftState.FINISHED_STATE,
+        ]
+
+        self._cursor.execute(sql, bindings, async_=True)
+        poll_state = self._cursor.poll()
+        state = poll_state.operationState
+
+        while state in STATE_PENDING:
+            logger.debug("Poll status: {}, sleeping".format(state))
+
+            poll_state = self._cursor.poll()
+            state = poll_state.operationState
+
+        # If an errorMessage is present, then raise a database exception
+        # with that exact message. If no errorMessage is present, the
+        # query did not necessarily succeed: check the state against the
+        # known successful states, raising an error if the query did not
+        # complete in a known good state. This can happen when queries are
+        # cancelled, for instance. The errorMessage will be None, but the
+        # state of the query will be "cancelled". By raising an exception
+        # here, we prevent dbt from showing a status of OK when the query
+        # has in fact failed.
+        if poll_state.errorMessage:
+            logger.debug("Poll response: {}".format(poll_state))
+            logger.debug("Poll status: {}".format(state))
+            dbt.exceptions.raise_database_error(poll_state.errorMessage)
+
+        elif state not in STATE_SUCCESS:
+            status_type = ThriftState._VALUES_TO_NAMES.get(state, 'Unknown')
```

Comment:
Maybe default to `'Unknown<{!r}>'.format(state)`, or something? It would be nice to know what is unknown if this happens!